Add validations to CheckinController

Bagurp opened the pull request:
* Changes 'completed' parameter in checkin from Nullable -> NotNull
* Update is only allowed if the check in is not completed unless made by admin
* Limit creation/read/updation to -
[] subject of the check in
[] the current PDL of the subject of the check in
[] admin
* Convert LocalDate to LocalDateTime for checkins service
* Add DB script to convert column type to time for Checkins table
* Issue - https://github.com/oci-labs/check-ins/issues/245

Bagurp commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInController.java:
```diff
@@ -47,17 +58,27 @@
-    public Set<CheckIn> findByValue(@Nullable UUID teamMemberId,
-                                     @Nullable UUID  pdlId, @Nullable Boolean completed) {          
-            return checkInservices.findByFields(teamMemberId, pdlId,completed);
+    public Set<CheckIn> findByValue(@Nullable UUID teamMemberId, @Nullable UUID  pdlId, @Nullable Boolean completed) {
```
These validations are expensive since DB call is made prior to validating the access of the current user. Could anyone think of a better way to do this?

meyerinj replied:
The @Secured annotation, when used on a method, can be used to specify a role.
https://micronaut-projects.github.io/micronaut-security/latest/guide/#secured

bhatiam12 replied:
that will not help in this situation, right ? Because we are validating if a current user is a PDL of the subject that he is doing a check in and not if he has a PDL role because many people will have PDL role but need to make sure they do the checkin for their team member.

meyerinj replied:
It will replace the securityService call.

As for the rest there is another option but whether it is better or not is debatable.

We can add claims to our user details mapper that will contain the user's PDL's ID and (if the user is a PDL) any team members for whom the user is a PDL. You would be able to check that against the request body instead of having to go through the user service.

mkimberlin replied:
No, it won't help us here. This validation has to be done by hand.

meyerinj replied:
I made a branch, [here](https://github.com/oci-labs/check-ins/tree/feature-245/also-with-claim-validation), which has a really hacky way to do that. If I understand what y'all are trying to do we should be able to use something like that.

mkimberlin replied:
I'm actually not against the idea of keeping a user's assigned team members in their user details...but I think we should make a separate story to investigate that.

Bagurp commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInController.java:
```diff
@@ -95,8 +131,15 @@
+        Boolean isAdmin = securityService.hasRole(RoleType.Constants.ADMIN_ROLE);
+        CheckIn checkin = checkInservices.read(id);
 
+        if(currentUser.getUuid().equals(checkin.getTeamMemberId()) || currentUser.getUuid().equals(checkin.getPdlId()) || isAdmin) {
```
Same here. Expensive validation

mkimberlin replied:
We need to allow the current PDL as well. There needs to be a case added for this.

Bagurp replied:
currentUser.getUuid().equals(checkin.getPdlId()) refers to current PDL.

Maroju100 commented:
POST - Checkin creation works even if member and PDL are not associated. 
Right now multiple checkins seems to be created (for multiple POSTs ) without any warning, not sure if that is expected behavior.
PUT-checkins looks to be getting updated my member even when submitted is true.
GET-Multiple checkins entries are shows since multiple POSTs are possible.

Bagurp commented:
@Maroju100 Good catch. I've added validations to impl class to both save/update. I also updated formatting. (Hide whitespace on github for better PR experience)

meyerinj commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInController.java:
```diff
@@ -38,6 +43,12 @@
     CheckInServices checkInservices ;
 
+    @Inject
+    CurrentUserServicesImpl currentUserServices;
```
Not specific to this story, but a better practice is to make these fields private and inject them via the constructor.

mkimberlin replied:
Also...why are you injecting the implementation rather than the interface here?

Bagurp replied:
good catch. I'll update it.

meyerinj commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInServicesImpl.java:
```diff
@@ -18,62 +20,80 @@
 
     @Inject
     private MemberProfileRepository memberRepo;
```
Private field injection is also bad. This can also be resolved by using constructor injection.

meyerinj replied:
I should add that the class must be annotated with @Singleton to use constructor injection

mkimberlin replied:
Yeah, I would change all your dependency fields to private and inject them via constructor.

meyerinj commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInServicesImpl.java:
```diff
@@ -18,62 +20,80 @@
         CheckIn checkInRet = null;
-        if(checkIn!=null){
+
+        if(checkIn!=null) {
```
Is there a scenario where we want the client to be able to attempt to save nothing and simply allow that to pass? Would it make more sense to use the @NotNull annotation and simply not allow this field to be null?

Bagurp replied:
Good catch. I'll make the required changes.

meyerinj commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInServicesImpl.java:
```diff
@@ -18,62 +20,80 @@
-        } else if(chkInDate.isBefore(LocalDate.EPOCH) || chkInDate.isAfter(LocalDate.MAX)) {
-            throw new CheckInBadArgException(String.format("Invalid date for checkin %s",memberId));
+
+        if(checkIn!=null) {
```
As with the save method would it make sense to simply not allow this to be null?

Bagurp replied:
I'll make the required changes.

meyerinj commented on server/src/main/java/com/objectcomputing/checkins/services/memberprofile/currentuser/CurrentUserServicesImpl.java:
```diff
@@ -24,4 +30,15 @@ public MemberProfile findOrSaveUser(@NotNull String name, @NotNull String workEm
 
     }
+
+    public MemberProfile currentUserDetails() {
```
When doing authentication it is probably easier to include a Principal argument in the controller than to use the security service.
https://micronaut-projects.github.io/micronaut-security/latest/guide/#retrievingAuthenticatedUser

mkimberlin commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInBadArgException.java:
```diff
@@ -1,11 +1,8 @@
-     */
-    private static final long serialVersionUID = 1L;
 
+    private static final long serialVersionUID = 1L;
```
I don't believe in this. You have offended my religion and we must now duel to the death. If you're not willing to face me, you may substitute a champion or generate a proper serialVersionUID.

Bagurp replied:
pardon, my lord - for I am just a humble developer cleaning unused comments. I choose to clean the code over facing your wrath ;)

Bagurp commented:
* If you are trying to run tests on IDE - run it with `MICRONAUT ENVIRONMENTS = test` so that it uses the right test config file.
* Please hide Whitespace on github for better PR review experience.

mkimberlin commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInServicesImpl.java:
```diff
@@ -1,86 +1,167 @@
+        } else if(!isAdmin) {
+            if(!currentUser.getId().equals(result.getTeamMemberId()) && !currentUser.getId().equals(result.getPdlId())) {
+                // Limit read to Subject of check-in, PDL of subject and Admin
+                throw new CheckInBadArgException(String.format("Member %s is unauthorized to do this operation", currentUser.getId()));
```
The ID isn't very meaningful to a user. I would nix the id here and just say something like "You are not authorized to perform this operation". It probably wouldn't hurt for us to have some AOP based logging that logs any thrown exceptions and records things like who the user was the method name and what the inputs to the method were.

mkimberlin replied:
To be clear, that wasn't intended to indicate you should implement AOP based logging in this PR.  Just a thought about something we might want to consider adding.

Bagurp replied:
Sounds good, I've created a new story to implement AOP based logging. For the current PR, I've updated error messages.

New Issue - https://github.com/oci-labs/check-ins/issues/309

mkimberlin commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInServicesImpl.java:
```diff
@@ -1,86 +1,167 @@
+            } else if(checkinRepo.findById(id).get().isCompleted()) {
+                // Update is only allowed if the check in is not completed unless made by admin
+                throw new CheckInBadArgException(String.format("Checkin with id %s is complete and cannot be updated", checkIn.getId()));
+            }
```
I feel like this could be made more concise by defining a method that would take care of the `if...throw` bit maybe?

It could take a lambda, a message, and an optional set of replacement values...or even a boolean instead of a lambda.  There's no need for the elses since an exception being thrown...so it could just be a sequence of validation calls. Just a thought.

Bagurp replied:
I've added a private method called validate in checkinServiceImpl in the following commit - https://github.com/oci-labs/check-ins/pull/283/commits/7437ac3545ee2026d01ec738582b3e7ee7c20a08

mkimberlin reviewed (approved):
couple of minor comments...nothing that I would block this proceeding though.

meyerinj commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInController.java:
```diff
@@ -33,10 +33,14 @@
 @Produces(MediaType.APPLICATION_JSON)
 @Consumes(MediaType.APPLICATION_JSON)
 @Tag(name="check-ins")
+@Singleton
```
Anything annotated with @Controller can also be treated as a Singleton so using both annotations is unnecessary.

meyerinj commented on server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInController.java:
```diff
@@ -33,10 +33,14 @@
 @Produces(MediaType.APPLICATION_JSON)
 @Consumes(MediaType.APPLICATION_JSON)
 @Tag(name="check-ins")
+@Singleton
```
Anything annotated with @controller can also be treated as a Singleton so using both annotations is unnecessary.

Bagurp replied:
Updated. I've removed @Singleton annotation from controller.
